### PR TITLE
Bump Nessie to 0.15.1 + related changes

### PR DESCRIPTION
**Pull request by snazy**
* Use new NessieApiV1
* Update reference syntax to `table-identifier ( '@' reference-name )? ( '#' pointer )?` (for Nessie-Iceberg-GC)
* Slightly nicer commit message from `NessieTableOperations`

Contributes to #3243

/cc @nastra

### CODE REVIEW

**Comment by snazy** [comment]
Moved to draft status due to https://github.com/projectnessie/nessie/issues/2312

**Comment by rdblue** [comment]
@nastra, can you take a look?

**Comment by rdblue** [comment]
This appears to change the relationship between Nessie and Iceberg metadata. Can you please explain what you're doing here and why it is needed? What is the model for Nessie and Iceberg tracking now? I thought it was that Nessie tracked different Iceberg metadata files. Is that no longer the case and Nessie is tracking structures within Iceberg?

**Comment by snazy** [comment]
@rdblue Sorry for the late reply. Yes, this one changes the relationship between Nessie and Iceberg metadata.

TL;DR the changes shall ensure that changes against the same table on different branches can later be merged together without having duplicate column-IDs or partition-IDs or the like.

You're right, initially (in the "early Nessie days"), every Nessie commit held a pointer to the table-metadata. This works fine until you reference the same table on different branches and perform e.g. schema changes (think: ALTER TABLE ADD COLUMN) on both branches, which leads to duplicate column-ids (in other words: same column id used for different columns), which then can lead to data corruption when branches get merged.

So the initial approach in this PR was to maintain table-metadata across all branches and "just" reference the snapshot-ID from in Nessie commits, which led to other issues (not explaining it here further, but schema changes became an issue again).

The current approach is more like the initial approach: have the pointer to table-metadata in Nessie commits but track state that's important across all Nessie branches (e.g. last-column-ID) globally. It's currently implemented via additional functionality in TableMetadata to retrieve the "global state" (last-column-ID, last-used-partition-ID, last-assigned-sequence-ID) as an object that's opaque to Nessie plus functionality to update a TableMetadata using that "global state".

**Comment by jackye1995** [comment]
It's still unclear to me how this would solve the merge conflict if you have 2 columns updated in different branches. For example, if I have branch `A` that added column `c1` as an int column, branch `B` also added column `c1` as a string column, even if they have different ID, it is unclear if the user's intention is to keep one `c1`  and which one to keep, or keep both (which is not possible due to duplicated column name).

**Review by nastra — COMMENTED** [review]
since Robert is out this week, I made some changes according to the review feedback

File: nessie/src/main/java/org/apache/iceberg/nessie/NessieTableOperations.java [hunk 1]
```diff
@@ -127,14 +155,32 @@ protected void doCommit(TableMetadata base, TableMetadata metadata) {
...
+        snapshot.snapshotId() != base.currentSnapshot().snapshotId())) {
+      commitMessage = String.format("Iceberg %s against %s", snapshot.operation(), tableName());
+      commitMeta.putProperties("iceberg.operation", snapshot.operation());
+    } else if (base != null && metadata.currentSchemaId() != base.currentSchemaId()) {
+      commitMessage = String.format("Iceberg schema change against %s", tableName());
+    } else {
+      commitMessage = String.format("Iceberg commit against %s", tableName());
+    }
+
+    return commitMeta.message(commitMessage);
```
[the remark this answers is not in the document]
Comment (rymurr) [reply nessie/src/main/java/org/apache/iceberg/nessie/NessieTableOperations.java:182]: I also find this rather confusing to read, some rather complicated booleans floating around. This seems like complexity in the name of removing code duplication

**Review by rymurr — CHANGES_REQUESTED** [review]
Added some comments, I think these changes could be architected better.

Also: this is a fairly fundamental change to how nessie and iceberg interact. I would expect more explanation of the change and its motivation and probably some docs changes (or links to nessie docs explaining iceberg/nessie interaction)

File: nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java [hunk 2]
```diff
@@ -338,21 +340,32 @@ String currentRefName() {
     return reference.getName();
   }
 
+  FileIO fileIO() {
```
[the remark this answers is not in the document]
Comment (nastra) [reply nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:343]: it's being used in 
```
private long snapshotIdFromMetadata(NessieCatalog catalog, String metadataLocation) {
    Snapshot snapshot = TableMetadataParser.read(catalog.fileIO(), metadataLocation).currentSnapshot();
    return snapshot != null ? snapshot.snapshotId() : -1;
  }
```

File: core/src/main/java/org/apache/iceberg/BaseMetastoreTableOperations.java [hunk 3]
```diff
@@ -184,13 +184,17 @@ protected void refreshFromMetadataLocation(String newLocation, Predicate<Excepti
...
 
-      this.currentMetadata = newMetadata.get();
+      this.currentMetadata = validateRefreshedMetadata(newMetadata.get());
       this.currentMetadataLocation = newLocation;
       this.version = parseVersion(newLocation);
     }
     this.shouldRefresh = false;
   }
 
+  protected TableMetadata validateRefreshedMetadata(TableMetadata metadata) {
```
[the remark this answers is not in the document]
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/BaseMetastoreTableOperations.java:194]: that would require making a bunch of fields `protected` in the base class. I renamed the method to `maybeValidateAndUpdateMetadata`. Hopefully that makes its purpose clearer

File: core/src/main/java/org/apache/iceberg/TableMetadata.java [hunk 4]
```diff
@@ -717,6 +717,19 @@ public TableMetadata removeSnapshotLogEntries(Set<Long> snapshotIds) {
         snapshots, newSnapshotLog, addPreviousFile(file, lastUpdatedMillis));
   }
 
+  public TableMetadata withUpdatedSnapshotAndSchema(long snapshotId, int schemaId, int sortOrderId, int specId) {
```
[the remark this answers is not in the document]
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/TableMetadata.java:720]: I think that things would be easier with the changes from https://github.com/apache/iceberg/pull/2957, where we could just call the builder and only update the actual fields that we need. Then we wouldn't need to have the `withUpdatedSnapshotAndSchema` method in `TableMetadata` and could do everything in the `NessieCatalog` where we need it. Maybe we should go forward with this as is right now and then refactor/improve this once #2957 is merged?

File: nessie/src/main/java/org/apache/iceberg/nessie/NessieTableOperations.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (nastra) [reply nessie/src/main/java/org/apache/iceberg/nessie/NessieTableOperations.java:182]: improved

File: core/src/main/java/org/apache/iceberg/TableMetadata.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (jackye1995) [reply core/src/main/java/org/apache/iceberg/TableMetadata.java:720]: For #2957 I was waiting for more clarity around the REST catalog to design the table metadata update builder semantics, so it probably won't be merged very soon. But I agree with Ryan that this seems like a very restrictive method, and the intention of the operation is unclear comparing to the other update methods (although I know how it fits in the Nessie context).

To update the schemaId, sortOrderId, specId to a specific version, I think it makes more sense to expose specific APIs in related update classes instead of having everything done through a separated method. What do you think?

File: core/src/main/java/org/apache/iceberg/BaseMetastoreTableOperations.java [hunk 5]
```diff
@@ -184,13 +184,17 @@ protected void refreshFromMetadataLocation(String newLocation, Predicate<Excepti
             "Table UUID does not match: current=%s != refreshed=%s", currentMetadata.uuid(), newUUID);
       }
 
-      this.currentMetadata = newMetadata.get();
+      this.currentMetadata = maybeValidateAndUpdateMetadata(newMetadata.get());
```
Comment (rymurr) [inline core/src/main/java/org/apache/iceberg/BaseMetastoreTableOperations.java:187]: I don't like this as a pattern. It feels like bad OOP. Can this be refactored? Or can `NessieCatalog` re-implement `refreshFromMetadataLocation`?

File: core/src/main/java/org/apache/iceberg/TableMetadata.java [hunk 6]
```diff
@@ -713,6 +713,64 @@ public TableMetadata removeSnapshotLogEntries(Set<Long> snapshotIds) {
...
 
+  /**
+   * Returns an updated {@link TableMetadata} with the current-snapshot-ID set to the given
+   * snapshot-ID and the snapshot-log reset to contain only the snapshot with the given snapshot-ID.
+   *
+   * @param snapshotId ID of a snapshot that must exist, or {@code -1L} to remove the current snapshot
+   *                   and return an empty snapshot log.
+   * @return {@link TableMetadata} with updated {@link #currentSnapshotId} and {@link #snapshotLog}
+   */
+  public TableMetadata withCurrentSnapshotOnly(long snapshotId) {
```
Comment (rymurr) [inline core/src/main/java/org/apache/iceberg/TableMetadata.java:724]: looking forward to getting #3580 in :-)

File: nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (rymurr) [reply nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:343]: in which class? Can't find it.

**Review by rymurr — APPROVED** [review]
LGTM thanks @snazy and @nastra

File: site/mkdocs.yml [hunk 7]
```diff
@@ -43,7 +43,7 @@ extra:
     Spark: /getting-started/
   versions:
     iceberg: 0.12.1
-    nessie: 0.9.2
+    nessie: 0.14.0
```
Comment (ajantha-bhat) [inline site/mkdocs.yml:46]:
```suggestion
    nessie: 0.15.1
```

File: core/src/main/java/org/apache/iceberg/BaseMetastoreTableOperations.java [hunk 8]
```diff
@@ -175,8 +182,7 @@ protected void refreshFromMetadataLocation(String newLocation, Predicate<Excepti
           .retry(numRetries).exponentialBackoff(100, 5000, 600000, 4.0 /* 100, 400, 1600, ... */)
           .throwFailureWhenFinished()
           .shouldRetryTest(shouldRetry)
-          .run(metadataLocation -> newMetadata.set(
-              TableMetadataParser.read(io(), metadataLocation)));
+          .run(metadataLocation -> newMetadata.set(metadataLoader.apply(metadataLocation)));
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseMetastoreTableOperations.java:185]: I'm just catching up on this PR, sorry to be late.

I'm curious why this uses a `Function` passed into refresh rather than defining a load metadata method. The function passed in from `NessieTableOperations` is actually a method reference, so couldn't this just be a call to that method, `loadTableMetadata`?

```java
    .run(metadataLocation -> loadTableMetadata(metadataLocation));

...

  public TableMetadata loadTableMetadata(String metadataLocation) {
    return TableMetadataParser.read(io(), metadataLocation);
  }
```
